onnxruntime/core/providers/qnn/qnn_execution_provider.cc

```diff
@@ -223,13 +239,49 @@
     }
   }
 
+  static const std::string QNN_DEVICE_ID = "device_id";
+  uint32_t device_id = 0;
+  auto dev_id_pos = provider_options_map.find(QNN_DEVICE_ID);
+  if (dev_id_pos != provider_options_map.end()) {
+    int value = std::stoi(dev_id_pos->second);
+    if (value < 0) {
+      LOGS_DEFAULT(WARNING) << "Invalid device ID '" << value
+                            << "', only >= 0 allowed. Set to " << device_id << ".";
+    } else {
+      device_id = static_cast<uint32_t>(value);
+    }
+  }
+
+  static const std::string QNN_HTP_ARCH = "htp_arch";
+  QnnHtpDevice_Arch_t htp_arch = QNN_HTP_DEVICE_ARCH_NONE;
+  auto htp_arch_pos = provider_options_map.find(QNN_HTP_ARCH);
+  if (htp_arch_pos != provider_options_map.end()) {
+    ParseHtpArchitecture(htp_arch_pos->second, htp_arch);
+  }
+
+  static const std::string QNN_SOC_MODEL = "soc_model";
+  uint32_t soc_model = QNN_SOC_MODEL_UNKNOWN;
+  auto soc_model_pos = provider_options_map.find(QNN_SOC_MODEL);
+  if (soc_model_pos != provider_options_map.end()) {
+    int value = std::stoi(soc_model_pos->second);
+    if (value < 0) {
```

Comment:
>if (value < 0) { [](start = 4, length = 16)

Could be a problem here since there's a limited number of values which are valid.

Reply:
QC has an enum (`Qnn_Soc_Model_t`) that lists some old SoC model values. However, that enum is now deprecated and QC will no longer add newer model numbers to it (probably because there are so many now). 

QC decided not to maintain a list of valid Soc model numbers, presumably because the valid set of model numbers can change independently of the QNN SDK version (I'm guessing). I'm inclined to follow their lead here and not maintain our own list of SoC model numbers. Instead, we can let the QNN backend apis validate the number for us.